README.md
# Static Markdown

A static web server with markdown rendering feature.

It serves regular HTTP content (HTML, JS, CSS, images, etc), but when you're browsing a ``.md``, ``.mdown`` or ``.markdown`` file, it's converted into HTML and rendered as a (hopefully) readable page.

![PyPI - Python Version](https://img.shields.io/pypi/pyversions/static-markdown)
![Travis CI status](https://img.shields.io/travis/com/brunobord/static-markdown)

## IMPORTANT WARNING

**This server is not recommended for production. It's a toy to serve static files and Markdown content, but there's no guarantee about security, performance, availability, etc.**

## Installation

### via PyPI

Install this tool using pip to download it [from PyPI](https://pypi.org/project/static-markdown/). You may prefer to use a virtualenv, but you may also want to install "user-wide".

```shell
pip install static-markdown
pip install --user static-markdown
```

### Using the source

We're advising you to use ``virtualenv`` to install this package. Clone this repository, and, inside your *virtualenv*, run the following:

```shell
pip install -e ./
```

## Usage

Once it's installed, you can just change your current shell session to the designated directory and run the following:

```shell
cd /path/to/directory
serve-md
```

Alternatively, you can also run this command line from anywhere, but set the target path as a command argument:

```shell
serve-md /path/to/directory
```

Using these default option, you can now browse your "static website" by pointing your browser to: <http://127.0.0.1:8080/>.

Stop the server with Ctrl-C.

### Options

```
usage: serve-md [-h] [-p PORT] [--markdown-template MARKDOWN_TEMPLATE]
                [--stylesheet STYLESHEET] [--version]
                [root]

positional arguments:
  root                  Path to serve statically

optional arguments:
  -h, --help            show this help message and exit
  -p PORT, --port PORT  Port number (0-65535)
  --markdown-template MARKDOWN_TEMPLATE
                        Path to an alternate HTML template for Markdown files
  --stylesheet STYLESHEET
                        Path to a custom stylesheet
  --version             Return version and exit
```

### Browsing

Let's consider the following path tree:

```
.
├── empty
├── images
│   └── knight.png
├── index.html
├── mdown
│   └── index.md
├── mdown-readme
│   └── README.md
└── subdir
    └── index.html
```

Here is a table describing the different pages rendered with the given URL in your browser

| URL                | Content                | Content type | Status Code           |
|:-------------------|:-----------------------|:-------------|:----------------------|
| /                  | index.html             | text/html    | 200 OK                |
| /empty             | directory listing      | text/html    | 200 OK                |
| /images/knight.png | knight.png (as binary) | image/png    | 200 OK                |
| /mdown/            | index.md (as HTML)     | text/html    | 200 OK                |
| /mdown/index.md    | index.md (as HTML)     | text/html    | 200 OK                |
| /mdown-readme      | README.md (as HTML)    | text/html    | 200 OK                |
| /subdir            | redirect to /subdir/   | -            | 301 MOVED PERMANENTLY |
| /subdir/           | subdir/index.html      | text/html    | 200 OK                |

### Indexes

When browsing a directory, ``serve-md`` will look after the following files, in that specific order: "index.html", "index.htm", "index.md", "README.md", "readme.md". The first one of them to be found will be served as the index page of the directory.

If none of them is found, ``serve-md`` will return a directory listing. If you're using the "Markdown template" feature, it'll be used when rendering this page.

### HTTP Verbs

GET & HEAD verbs are implemented.

POST, PUT, DELETE will result into a 405 Method Not Allowed error.

### Markdown templates

By default, Markdown files will be rendered as HTML using a minimal CSS. Using the ``--markdown-template`` option, you can use your own HTML template with a custom CSS to render generate the ``text/html``. You can find various examples in the ``example-options`` directory in the source repository.

We've used various minimalist CSS frameworks for rendering and layout:

* [miligram](https://milligram.io/),
* [mini css](https://minicss.org/),
* [Github Markdown CSS](https://github.com/sindresorhus/github-markdown-css),

### Custom styles

If the only thing you need is to customize styles, but not the structure of your HTML page, you can use the ``--stylesheet`` option to point at a CSS file.

Example:

```sh
serve-md /path/to/directory --stylesheet ~/my-style.css
```

## Hack

1. Clone this repository.
2. Write tests.
3. Implement stuff.
4. Open a Pull Request.

**Important:** Each time you're modifying any line of the server's code, if you want to test your patch **live**, you **HAVE** to stop and restart the application.

### Run automated tests

In a shell session, run tests with:

```shell
make test
```


CHANGELOG.md
# Changelog

## master (unreleased)

Nothing here yet.

## v0.4.0 (2021-05-10)

### New features

* Add/Confirm Python 3.8 compatibility (#10).
* Add/Confirm Python 3.9 compatibility (#11).

### Other improvements & fixes

* Fixed CI & solved regressions by upgrading of `pytest-xprocess` & `Markdown` libraries.
* Added the compatibility badges - courtesy of shields.io (#15).
* Added the Travis CI badge - also switched from travis-ci.org to the .com instance, hopefully more future-proof (#17).

## v0.3.0 (2019-10-05)

### New features

* Added HEAD implementation (#7).
* Document and make sure that the POST, PUT, DELETE and PATCH verbs would return a 405 error.
* Enable using a custom stylesheet.

### Other improvements

* Default style: limit ``<img>`` width up to the page width.

## v0.2.0 (2019-06-19)

### New features

* Enabled directory listing.
* Redirect to URL with trailing slash when pointing at a directory (#4).
* A better logging using loguru.
* A better way to test the application, without having to start the server in another shell.
* Implement version handling via a ``__version__`` module property.
* Added server name and version in the returned Server headers (#6).

### Fixes & other changes

* Fix 404 error when trying to browse file with spaces in their names (#1).
* Readme instructions to install from PyPI.
* Document the fact that the server is not production-ready (#5).
* Added twine in extra requirements (dev).
* Added travis support for CI.

## v0.1.0 (2019-06-13)

Initial release.

* Statically serve your local directory,
* Render Markdown as HTML,
* Searching for an index (HTML or Markdown) when trying to browse a directory,
* Customize Markdown HTML rendering using a custom HTML template.


tests/conftest.py
from os.path import abspath, dirname, join

import pytest
from xprocess import ProcessStarter

current_dir = dirname(__file__)
project_dir = abspath(join(current_dir, ".."))
executable_dir = abspath(join(project_dir, "static_markdown"))
example_dir = abspath(join(project_dir, "example"))


@pytest.fixture
def http_server(xprocess):
    class Starter(ProcessStarter):
        pattern = "Serving"
        args = ["python", f"{executable_dir}/cli.py", example_dir]

    _ = xprocess.ensure("server-default", Starter)
    info = xprocess.getinfo("server-default")
    assert info.isrunning()
    yield info
    xprocess.getinfo("server-default").terminate()


@pytest.fixture
def http_server_custom_style(xprocess):
    class Starter(ProcessStarter):
        pattern = "Serving"
        args = [
            "python",
            f"{executable_dir}/cli.py",
            example_dir,
            "--stylesheet",
            f"{example_dir}/style.css",
            "--port",
            "8081",
        ]

    _ = xprocess.ensure("server-custom-css", Starter)
    info = xprocess.getinfo("server-custom-css")
    assert info.isrunning()
    yield info
    xprocess.getinfo("server-custom-css").terminate()


tests/test_cli.py
from unittest.mock import MagicMock, patch

from static_markdown.server import main


def test_main_regular_call():
    with patch("argparse.ArgumentParser.parse_args") as argument_mock:
        argument_mock.return_value = MagicMock(version=False, root=".", port=9999)
        with patch("http.server.HTTPServer.serve_forever") as serve_mock:
            serve_mock.return_value = True
            main()
            serve_mock.assert_called_once()


def test_main_with_version():
    with patch("argparse.ArgumentParser.parse_args") as argument_mock:
        argument_mock.return_value = MagicMock(version=True, root=".", port=9999)
        with patch("http.server.HTTPServer.serve_forever") as serve_mock:
            serve_mock.return_value = True
            main()
            serve_mock.assert_not_called()


tests/test_index.py
import requests


def test_home(http_server):
    response = requests.get("http://127.0.0.1:8080/index.html")
    assert response.status_code == 200
    index_html = response.content
    response = requests.get("http://127.0.0.1:8080/")
    assert response.status_code == 200
    assert response.content == index_html


def test_subdir(http_server):
    response = requests.get("http://127.0.0.1:8080/subdir/index.html")
    assert response.status_code == 200
    index_html = response.content
    response = requests.get("http://127.0.0.1:8080/subdir/")
    assert response.status_code == 200
    assert response.content == index_html


def test_alternate_index(http_server):
    response = requests.get("http://127.0.0.1:8080/alternate-index/")
    assert response.status_code == 200
    index_html = response.content
    response = requests.get("http://127.0.0.1:8080/alternate-index/index.htm")
    assert response.status_code == 200
    assert response.content == index_html
    assert response.content == b"I am the alternate index\n"


def test_empty_404(http_server):
    response = requests.get("http://127.0.0.1:8080/empty/")
    assert response.status_code == 200


def test_dirlist(http_server):
    response = requests.get("http://127.0.0.1:8080/no-index/")
    assert response.status_code == 200
    content = response.content
    # Another Directory
    assert b'<a href="/no-index/subdir/">subdir/</a>' in content
    # Two files
    assert b'<a href="/no-index/file.html">file.html</a>' in content
    assert b'<a href="/no-index/other-file.html">other-file.html</a>' in content


static_markdown/helpers.py
import email
import time
import urllib.parse
from contextlib import contextmanager
from os import chdir, getcwd

DEFAULT_MARKDOWN_TEMPLATE = """<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8" />
<meta http-equiv="X-UA-Compatible" content="chrome=1">
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<title>{title}</title>
<style>
{style}
</style>
</head>
<body>
    <main>{content}</main>
</body>
</html>
"""

DEFAULT_MARKDOWN_STYLE = """
body { font-family: Helvetica, arial, freesans, clean, sans-serif; }
main { max-width: 70ch; padding: 2ch; margin: auto; }
code { font-family: monospace; }
main img { max-width: 70ch; }
"""


@contextmanager
def cd(dirname):
    """
    Context manager to temporarily change current directory.
    """
    _curdir = getcwd()
    chdir(dirname)
    yield
    chdir(_curdir)


def date_time_string(timestamp=None):
    """Return the current date and time formatted for a message header."""
    if timestamp is None:
        timestamp = time.time()
    return email.utils.formatdate(timestamp, usegmt=True)


def normalize_path(path):
    """
    Return URL-entity-free pathname
    """
    try:
        path = urllib.parse.unquote(path, errors="surrogatepass")
    except UnicodeDecodeError:
        path = urllib.parse.unquote(path)
    return path


static_markdown/document.py
import mimetypes
import os
from os.path import basename, isdir, isfile, join, splitext

import markdown
from markdown.extensions.toc import TocExtension
from mdx_gfm import GithubFlavoredMarkdownExtension
from slugify import slugify

from .helpers import date_time_string, normalize_path


def convert_md_source(source):
    "Convert Markdown content into HTML"
    html = markdown.markdown(
        source,
        extensions=[
            GithubFlavoredMarkdownExtension(),
            TocExtension(permalink=False, slugify=slugify),
        ],
    )
    return html


class DocumentError(Exception):
    """
    Error on the document
    """

    def __init__(self, status_code, message):
        self.status_code = status_code
        self.message = message


class RedirectionException(Exception):
    """
    Redirect to location
    """

    def __init__(self, location, status_code=301):
        self.location = location
        self.status_code = status_code


class Document(object):
    def __init__(self, root, path, markdown_template=None, style=None):
        # Init
        self.content_length = 0
        self.last_modified = None

        # Seek and load
        self.root = root
        self.path = normalize_path(path)
        self.markdown_template = markdown_template
        self.style = style
        self.filepath = self.get_filepath()
        self._type = self.guess_type()
        self.content = self.get_content()

    def guess_type(self):
        """
        Guess and return type
        """
        # split extension
        _, extension = splitext(self.filepath)
        if extension in (".md", ".markdown", ".mdown"):
            _type = "text/markdown"
        else:
            _type, encoding = mimetypes.guess_type(self.filepath, strict=False)
        return _type or "text/plain"

    def find_index(self, path):
        """
        Find the proper index in the directory
        """
        indexes = ("index.html", "index.htm", "index.md", "README.md", "readme.md")
        temp_path = self.root + path
        # This is a directory, but it does not end with "/"
        if isdir(temp_path) and not path.endswith("/"):
            raise RedirectionException(self.path + "/")

        if path == "/" or isdir(temp_path):
            for index in indexes:
                index_path = join(temp_path, index)
                if isfile(index_path):
                    return index_path

    def get_filepath(self):
        """
        Return the filepath on the filesystem.
        """
        index = self.find_index(self.path)
        if index:
            return index

        filepath = self.root + self.path
        if isfile(filepath) or isdir(filepath):
            return filepath
        # I don't know what it should be here...
        raise DocumentError(404, "File Not Found")

    def get_dirlist(self):
        content = []
        content.append(f"# Directory listing for {self.path}")
        path = self.path
        if not path.endswith("/"):
            path = path + "/"
        for name in os.listdir(self.filepath):
            fullpath = join(self.filepath, name)
            if isdir(fullpath):
                content.append(f"* [{name}/]({path}{name}/)")
            if isfile(fullpath):
                content.append(f"* [{name}]({path}{name})")
        content = "\n".join(content)
        return self.render_markdown(content)

    def render_markdown(self, mdown_content):
        # Markdown to HTML
        content = convert_md_source(mdown_content)
        # Render in template
        content = self.markdown_template.format(
            title=basename(self.filepath), style=self.style, content=content
        )
        self.content_length = len(content)
        self.last_modified = date_time_string(os.path.getmtime(self.filepath))
        # Somehow change the type, since we're rendering this as HTML
        self._type = "text/html"
        return content.encode()

    def get_content_markdown(self):
        """
        Load Markdown content and render it
        """
        with open(self.filepath, "r") as fh:
            mdown_content = fh.read()
        return self.render_markdown(mdown_content)

    def get_content(self):
        """
        Return the (binary) content to stream to the HTTP client.
        """
        if self._type == "text/markdown":
            return self.get_content_markdown()

        if isdir(self.filepath):
            return self.get_dirlist()

        try:
            fh = open(self.filepath, "rb")
            content = fh.read()
        except OSError:
            raise DocumentError(404, "File Not Found or I/O Error")
        finally:
            # Close file handler
            fh.close()
        # retrieve content length & last modified time.
        self.content_length = len(content)
        self.last_modified = date_time_string(os.path.getmtime(self.filepath))
        # Return file content
        return content


static_markdown/server.py
#!/usr/bin/env python3
import argparse
import sys
from http import HTTPStatus
from http.server import BaseHTTPRequestHandler, HTTPServer
from os.path import abspath, isdir

from loguru import logger

from . import __version__
from .document import Document, DocumentError, RedirectionException
from .helpers import DEFAULT_MARKDOWN_STYLE, DEFAULT_MARKDOWN_TEMPLATE, cd

log_format = (
    "<green>{time:YYYY-MM-DD HH:mm:ss.SSS}</green> "
    "| <level>{level: <8}</level> | <level>{message}</level>"
)

logger.configure(
    handlers=[{"sink": sys.stderr, "format": log_format}],
    extra={"reqid": "-", "ip": "-", "user": "-"},
)


class Server(HTTPServer):
    def __init__(self, *args, **kwargs):
        self._root = kwargs.pop("root", ".")
        self._root_url = kwargs.pop("root_url", ".")
        stylesheet = kwargs.pop("stylesheet", None)

        # Default style
        self._style = DEFAULT_MARKDOWN_STYLE
        if stylesheet:
            self._style = stylesheet.read()

        # Template Handling
        _template_handler = kwargs.pop("markdown_template", None)
        if _template_handler:
            self._markdown_template = _template_handler.read()
        else:
            self._markdown_template = DEFAULT_MARKDOWN_TEMPLATE
        super().__init__(*args, **kwargs)


class StaticMarkdownHandler(BaseHTTPRequestHandler):
    server_version = "Static-Markdown/" + __version__

    @property
    def root(self):
        if not hasattr(self, "_root"):
            self._root = getattr(self.server, "_root")
        return self._root

    @property
    def markdown_template(self):
        if not hasattr(self, "_markdown_template"):
            self._markdown_template = getattr(self.server, "_markdown_template")
        return self._markdown_template

    @property
    def style(self):
        if not hasattr(self, "_style"):
            self._style = getattr(self.server, "_style")
        return self._style

    def get_document(self):
        """
        Return the appropriate document corresponding to the path.
        """
        try:
            document = Document(
                self.root, self.path, self.markdown_template, self.style
            )
        except DocumentError as exc:
            self.send_error(exc.status_code, exc.message)
            return
        except RedirectionException as exc:
            self.send_response(HTTPStatus.MOVED_PERMANENTLY)
            self.send_header("Location", f"{exc.location}")
            self.end_headers()
            return

        # Before sending the headers, send the response status
        self.send_response(HTTPStatus.OK)
        self.send_header("Content-Length", str(document.content_length))
        self.send_header("Content-type", document._type)
        self.send_header("Last-Modified", document.last_modified)
        # Ending headers here
        self.end_headers()
        return document

    def do_HEAD(self):
        """
        Handle HEAD requests
        """
        self.get_document()

    def do_GET(self):
        """
        Handle GET requests
        """
        document = self.get_document()
        if document:
            self.wfile.write(document.content)

    def _do_405(self):
        """
        Handle POST/PUT/PATCH/DELETE requests. Return 405.
        """
        self.send_response(HTTPStatus.METHOD_NOT_ALLOWED)
        self.end_headers()
        return

    do_PUT = do_DELETE = do_PATCH = do_POST = _do_405

    def log_message(self, format, *args):
        """
        Log handling
        """
        logger.info("%s - %s" % (self.address_string(), format % args))

    def log_error(self, format, *args):
        """
        Error log handling
        """
        logger.error("%s - %s" % (self.address_string(), format % args))

    def version_string(self):
        """
        Return the server software version string.
        """
        return self.server_version


def serve(root, port=8080, markdown_template=None, scheme="http", stylesheet=None):
    root = abspath(root)
    if not isdir(root):
        logger.error("Error: `{}` is not a directory".format(root))
        return

    root_url = f"{scheme}://127.0.0.1:{port}"
    logger.info(f"Serving `{root}`..." f"\nGo to: {root_url}")
    with cd(root):
        server_address = ("", port)
        httpd = Server(
            server_address,
            StaticMarkdownHandler,
            root=root,
            markdown_template=markdown_template,
            root_url=root_url,
            stylesheet=stylesheet,
        )
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            pass
    logger.info("Bye...")


def port(number):
    """
    A port value is a number between 0 and 65535.
    """
    try:
        number = int(number)
    except ValueError:
        raise argparse.ArgumentTypeError("invalid int value: '{}'".format(number))
    if 0 <= number <= 65535:
        return number
    raise argparse.ArgumentTypeError("port must be 0-65535")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("root", nargs="?", default=".", help="Path to serve statically")
    parser.add_argument(
        "-p", "--port", type=port, default=8080, help="Port number (0-65535)"
    )
    parser.add_argument(
        "--markdown-template",
        default=None,
        type=argparse.FileType("r"),
        help="Path to an alternate HTML template for Markdown files",
    )
    parser.add_argument(
        "--stylesheet",
        default=None,
        type=argparse.FileType("r"),
        help="Path to a custom stylesheet",
    )
    parser.add_argument(
        "--version", action="store_true", default=False, help="Return version and exit"
    )
    args = parser.parse_args()
    if args.version:
        print(f"Static Markdown v{__version__}")
        return
    serve(
        root=args.root,
        port=args.port,
        markdown_template=args.markdown_template,
        stylesheet=args.stylesheet,
    )
